### PR TITLE
Added improvements to category url key validation logic

### LINKED ISSUE #13689: Cannot create catagory's name with thai langauge
### Preconditions
1. Magento 2.2.2
2. PHP 7.1.11

### Steps to reproduce
1. Create new sub category
2. Use thai language in name field 

### Expected result
1. get new category with thai name language Or have a meaningful message that we need to fix url key manually

### Actual result

1. get error in picture
screen shot 2561-02-16 at 3 28 43 pm
And message not clear at all

### PR DESCRIPTION
**Pull request by sergiy-v**
### Description
Improved validate URL key logic for category.
Screen Shot 2020-03-24 at 11 20 15 AM
Screen Shot 2020-03-24 at 11 21 30 AM

I have added improvements to validate URL key logic for category instead of adding a solution for Thai language because of:
- `\Magento\Framework\Filter\Translit` can not provide translation for Thai language (the convert table does not contain Thai symbols). It will be hard to add and manage all needed symbols for all languages to the convert table.
- Magento Community Engineering provides on localization extensions like [Japanese Localization](https://marketplace.magento.com/community-engineering-japan-common.html) which provide a solution for such cases (has plugin to filter value in other way).

### Fixed Issues
#13689

### Manual testing scenarios
See #13689 issue description for details.

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @sergiy-v,
Your change improves message, however you still will not be able to create category -> it's not fixing the listed issue.

Could you update your changes?

**Comment by sergiy-v** [comment]
> Hi @sergiy-v,
> Your change improves message, however you still will not be able to create category -> it's not fixing the listed issue.
> 
> Could you update your changes?

Hello @ihor-sviziev,
Yes, it's not fixed the issue (not providing translation) as I mentioned in description. I have added the reasons why the logic it's should not be changed.
I'll be happy to add a solution, but as discussed in related tasks some expectation can be wrong, the other changes can be incompatible due to backward compatibility or so.
That's why I suggest to fix the issue by providing more details to the admin user in error message to solve the issue.
@ihor-sviziev please suggest the possible solution and I'll try to add it to PR.

Thank you.

**Comment by ihor-sviziev** [comment]
@sergiy-v i got your point, seems reasonable. Will review your PR little bit later

File: dev/tests/integration/testsuite/Magento/Catalog/Controller/Adminhtml/CategoryTest.php [hunk 1]
```diff
@@ -890,4 +890,70 @@ public function testSaveWithCustomBackendNameAction(): void
...
+    }
+
+    /**
+     * Verify that the category cannot be saved if category URL key can not be converted to Latin (like Thai)
+     *
+     * @return void
+     */
+    public function testSaveWithThaiCategoryUrlKeyAction(): void
+    {
+        $categoryPath = 'ประเภท';
```
Comment (ihor-sviziev) [inline dev/tests/integration/testsuite/Magento/Catalog/Controller/Adminhtml/CategoryTest.php]: That’s not actually category path. So would be great to rename it 
```suggestion
        $categoryUrlKey = 'ประเภท';
```

File: app/code/Magento/CatalogUrlRewrite/Observer/CategoryUrlPathAutogeneratorObserver.php [hunk 2]
```diff
@@ -107,33 +111,70 @@ public function execute(\Magento\Framework\Event\Observer $observer)
...
+    /**
+     * Validate URL key value
+     *
+     * @param Category $category
+     * @param string $urlKey
+     *
+     * @throws LocalizedException
+     * @throws LogicException
+     */
+    private function validateUrlKey($category, $urlKey)
```
Comment (ihor-sviziev) [inline app/code/Magento/CatalogUrlRewrite/Observer/CategoryUrlPathAutogeneratorObserver.php]: Could you add argument and return types for this method?

Comment (ihor-sviziev) [inline app/code/Magento/CatalogUrlRewrite/Observer/CategoryUrlPathAutogeneratorObserver.php]: Could you add argument and return types for this method?
And the same for updateUrlKey method

File: dev/tests/integration/testsuite/Magento/Catalog/Controller/Adminhtml/CategoryTest.php (hunk 1 above)
Comment (ihor-sviziev) [inline dev/tests/integration/testsuite/Magento/Catalog/Controller/Adminhtml/CategoryTest.php]: This is not a path, but url key

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @sergiy-v,
Please review my comments and static test failures (except copy paste failure, this is known issue) and update your changes accordingly

File: dev/tests/integration/testsuite/Magento/Catalog/Controller/Adminhtml/CategoryTest.php (hunk 1 above)
Comment (sergiy-v) [reply dev/tests/integration/testsuite/Magento/Catalog/Controller/Adminhtml/CategoryTest.php]: Nice catch! Changed, thank you.

File: app/code/Magento/CatalogUrlRewrite/Observer/CategoryUrlPathAutogeneratorObserver.php (hunk 2 above)
Comment (sergiy-v) [reply app/code/Magento/CatalogUrlRewrite/Observer/CategoryUrlPathAutogeneratorObserver.php]: The changes has been added, thanks.

**Comment by sergiy-v** [comment]
> Hi @sergiy-v,
> Please review my comments and static test failures (except copy paste failure, this is known issue) and update your changes accordingly

Thanks for suggestions, the changes has been added. I believe the tests errors not related to the changes, please check it.

Thank you.

**Comment by ihor-sviziev** [comment]
Hi @lenaorobei @slavvka,
could you give some more info why dB compare test fails, all results inside are green. Tried restarting, but the same result

**Comment by ihor-sviziev** [comment]
Oh, looks after merging 2.4 develop the issue is gone. Thank you @lenaorobei !

**Review by ihor-sviziev — APPROVED** [review]
Looks like test failures are not related to changes from this PR

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7208](https://jira.corp.magento.com/browse/ENGCOM-7208) has been created to process this Pull Request

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed 

**Before:**
Screenshot from 2020-03-26 12-11-01

**After:**
Screenshot from 2020-03-26 12-19-06

Screenshot from 2020-03-26 12-30-14

**Comment by engcom-Echo** [comment]
Failed functional tests not related to the changes in this PR - [comment](https://github.com/magento/magento2/pull/27412#pullrequestreview-381736009)

**Pull request merged**